torch/utils/tensorboard/embedding.py

```diff
@@ -0,0 +1,67 @@
+import os
+
+
+def make_tsv(metadata, save_path, metadata_header=None):
+    if not metadata_header:
+        metadata = [str(x) for x in metadata]
+    else:
+        assert len(metadata_header) == len(metadata[0]), \
+            'len of header must be equal to the number of columns in metadata'
+        metadata = ['\t'.join(str(e) for e in l)
+                    for l in [metadata_header] + metadata]
+    import sys
+    if sys.version_info[0] == 3:
+        with open(os.path.join(save_path, 'metadata.tsv'), 'w', encoding='utf8') as f:
+            for x in metadata:
+                f.write(x + '\n')
+    else:
+        with open(os.path.join(save_path, 'metadata.tsv'), 'wb') as f:
+            for x in metadata:
+                f.write((x + '\n').encode('utf-8'))
+
+
+# https://github.com/tensorflow/tensorboard/issues/44 image label will be squared
+def make_sprite(label_img, save_path):
+    import math
+    import numpy as np
+    from .x2num import make_np
+    from .utils import make_grid
+    from PIL import Image
+    # this ensures the sprite image has correct dimension as described in
+    # https://www.tensorflow.org/get_started/embedding_viz
+    nrow = int(math.ceil((label_img.size(0)) ** 0.5))
+    arranged_img_CHW = make_grid(make_np(label_img), ncols=nrow)
+
+    # augment images so that #images equals nrow*nrow
+    arranged_augment_square_HWC = np.ndarray((arranged_img_CHW.shape[2], arranged_img_CHW.shape[2], 3))
+    arranged_img_HWC = arranged_img_CHW.transpose(1, 2, 0)  # chw -> hwc
+    arranged_augment_square_HWC[:arranged_img_HWC.shape[0], :, :] = arranged_img_HWC
+    im = Image.fromarray(np.uint8((arranged_augment_square_HWC * 255).clip(0, 255)))
+    im.save(os.path.join(save_path, 'sprite.png'))
+
+
+def append_pbtxt(metadata, label_img, save_path, subdir, global_step, tag):
+    from posixpath import join
```

Comment:
Please move this import to the top of the file.

Reply:
Actually, why can't `os.path.join()` be used instead?

Reply:
Windows compatibility: https://github.com/lanpa/tensorboardX/commit/29c6194f10a75ebb76ecf6e5c94e7598c3cafcb0#diff-026003bb872ddb8769e4ec04567043e4